mitodl/mitxonline
# Add queries to migrate order, line and transaction from MicroMaster

#### Pre-Flight checklist
#### What are the relevant tickets?
https://github.com/mitodl/mitxonline/issues/657

#### What's this PR do?

- Adding a new state 'partially_refunded' to bring over partially refunded orders from MM
- Making 'reference_number' as readly-only field in django admin (This field can't be manually modified as we rely on it for migration)
- Adding migration queries to bring over orders/lines/transactions from MicroMaster

#### How should this be manually tested?

- Run the following query to check if there are corresponding products for the course runs in the local DB
   (If any course run doesn't have a product, then add them from Django admin)
  
```
     SELECT * FROM courses_courserun AS run 
     LEFT JOIN public.ecommerce_product AS product 
       ON run.id = product.object_id AND product.content_type_id = 60
    WHERE product.id is NULL
```
 - Create a local foreign data connection named 'micromasters'  (Only need this step when testing locally)
 ```
CREATE EXTENSION IF NOT EXISTS postgres_fdw;

CREATE SERVER IF NOT EXISTS micromaster_bridge
	FOREIGN DATA WRAPPER postgres_fdw
    OPTIONS (host 'micromasters-rc-db.cbnm7ajau6mi.us-east-1.rds.amazonaws.com', dbname 'micromasters', port '15432');

CREATE USER MAPPING IF NOT EXISTS FOR postgres
	SERVER micromaster_bridge
    OPTIONS (user 'micromasters', password '------');

DROP SCHEMA IF EXISTS micromasters CASCADE;
CREATE SCHEMA micromasters;
IMPORT FOREIGN SCHEMA public 
FROM SERVER micromaster_bridge INTO micromasters;
```

- Run the migration script
```
docker-compose run --rm web ./manage.py import_micromasters_data
```

- Verify orders/line/transactions are imported correctly (compare with MicroMaster RC using reference_number)

Orders
   
![image](https://user-images.githubusercontent.com/3138890/184992866-f33621b6-70ce-47a0-ab93-089b592c494d.png)

Line:

![image](https://user-images.githubusercontent.com/3138890/184993060-ed9a29ea-2a04-460f-a995-de7ca3e7b874.png)

Transaction
<img width="808" alt="image" src="https://user-images.githubusercontent.com/3138890/186981827-c034e06d-ec0e-4284-a7ea-f3c3c582f0c2.png">

Receipt on frontend
<img width="1117" alt="image" src="https://user-images.githubusercontent.com/3138890/186982225-6268fd1e-e9fb-4e28-bc7a-68160f546c1a.png">
